fonts/generate.py

```diff
@@ -0,0 +1,657 @@
+#define SMUFL_COUNT {len_smufl_codes}
+
+}} // namespace vrv
+
+#endif
+"""
+
+FONTFORGE_SCRIPT = (
+    """import os; fontforge.open("{input_fontpath}").generate("{output_fontpath}")"""
+)
+
+FONTFACE_WRAPPER = """@font-face {{
+    font-family: '{fontname}';
+    src: url(data:application/font-woff2;charset=utf-8;base64,{b64encoding}) format('woff2');
+    font-weight: normal;
+    font-style: normal;
+}}"""
+
+
+B64_FONT_LICENSE = """
+This font is licensed under the SIL Open Font License, http://scripts.sil.org/OFL, with 
+Reserved Font Name "{fontname}".
+
+See https://github.com/rism-digital/verovio/blob/develop/fonts/README.md for more information.
+
+This file is a subset of the full font containing only the glyphs supported by Verovio for efficiency. 
+Please do not use directly.
+"""
+
+log = logging.getLogger(__name__)
+
+
+def generate_smufl(opts: Namespace) -> bool:
+    """
+    Generates the `smufl.h` file for Verovio.
+
+    :param opts: A set of options from the argument parser sub-command.
+    :return: True if successful, False otherwise.
+    """
+    supported_glyphs_fname: str = opts.supported
+    header_out_dir: str = opts.header_out
+    supported_path: Path = Path(supported_glyphs_fname)
+    header_out_path: Path = Path(header_out_dir)
+    header_file_path: Path = Path(header_out_path, "smufl.h")
+
+    if not supported_path.is_file() or not os.access(supported_path, os.R_OK):
+        log.error(f"Could not find or read {supported_glyphs_fname} file.")
+        return False
+
+    if not os.access(header_out_path, os.W_OK):
+        log.error(f"Could not write to {header_out_dir}.")
+        return False
+
+    log.debug("SMuFL header will be stored in %s", header_out_path)
+    supported_glyphs: dict = __get_supported_glyph_codes(supported_glyphs_fname)
+
+    fmt_supported_glyphs: list = [
+        f"    SMUFL_{gcode}_{gname} = 0x{gcode}," for gcode, gname in supported_glyphs.items()
+    ]
+
+    log.debug("SMuFL header will contain %s supported glyphs", len(fmt_supported_glyphs))
+    fmt_glyph_list = "\n".join(fmt_supported_glyphs)
+    fmt_header = SMUFL_HEADER.format(
+        smufl_glyph_list=fmt_glyph_list, len_smufl_codes=len(fmt_supported_glyphs)
+    )
+
+    log.debug("Writing %s", header_file_path)
+    with open(header_file_path, "w") as header_inc:
+        header_inc.write(fmt_header)
+
+    log.debug("Finished writing SMuFL header")
+    return True
+
+
+def extract_fonts(opts: Namespace) -> bool:
+    """
+    Takes a font file and extracts the necessary files for Verovio. Generates a glyph file for each glyph, and an SVG
+    glyph file containing only the supported SMuFL glyphs.
+
+    :param opts: A set of options from the argument parser sub-command.
+    :return: True if successful, False otherwise.
+    """
+    source_dir: str = opts.source
+    fontname: str = opts.fontname
+    data_dir: str = opts.data
+
+    source_pth: Path = Path(source_dir)
```

Comment:
Is there a specific reason for abbreviating "path" to "pth"? I think one additional character wouldn't harm.

Reply:
It was an attempt to avoid [a Python anti-pattern](https://docs.quantifiedcode.com/python-anti-patterns/readability/putting_type_information_in_a_variable_name.html), while at the same time trying to make sure I'm keeping all the different variations of that data separate and well-named, and also trying to keep some semblance of the old code so that I could easily find my way back to it while I was porting it over. (The XML processors, for example, don't take `Path` objects, so I have to use the string variant on them...)